src/NServiceBus.Core.Tests/Routing/DetermineRoutingForMessageBehaviorTests.cs

```diff
@@ -0,0 +1,136 @@
+﻿namespace NServiceBus.Core.Tests.Routing
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using NServiceBus.Pipeline.Contexts;
+    using NServiceBus.Routing;
+    using NServiceBus.Transports;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class DetermineRoutingForMessageBehaviorTests
+    {
+        [Test]
+        public void Should_use_explicit_route_for_sends_if_present()
+        {
+            var behavior = InitializeBehavior();
+            var options = new SendOptions();
+
+            options.SetDestination("destination endpoint");
+
+            var context = new OutgoingContext(null, typeof(MyMessage), null, options);
+
+            behavior.Invoke(context, () => { });
+
+            var routingStrategy = (DirectToTargetDestination)context.Get<RoutingStrategy>();
+
+            Assert.AreEqual("destination endpoint", routingStrategy.Destination);
+        }
+
+        [Test]
+        public void Should_route_to_local_endpoint_if_requested_so()
+        {
+             var behavior = InitializeBehavior("MyLocalAddress");
+            var options = new SendOptions();
+
+            options.RouteToLocalEndpointInstance();
+
+            var context = new OutgoingContext(null, typeof(MyMessage), null, options);
+
+            behavior.Invoke(context, () => { });
+
+            var routingStrategy = (DirectToTargetDestination)context.Get<RoutingStrategy>();
+
+
+            Assert.AreEqual("MyLocalAddress", routingStrategy.Destination);
+        }
+
+        [Test]
+        public void Should_route_using_the_mappings_if_no_destination_is_set()
+        {
+            var router = new FakeRouter();
+
+            var behavior = InitializeBehavior(router: router);
+            var options = new SendOptions();
+
+            var context = new OutgoingContext(null, typeof(MyMessage), null, options);
+
+            behavior.Invoke(context, () => { });
+
+            var routingStrategy = (DirectToTargetDestination)context.Get<RoutingStrategy>();
+
+            Assert.AreEqual("MappedDestination", routingStrategy.Destination);
+        }
+
+        [Test]
+        public void Should_use_to_all_subscribers_strategy_for_events()
+        {
+            var behavior = InitializeBehavior();
+            var options = new PublishOptions();
+
+            var context = new OutgoingContext(null, typeof(MyMessage), null, options);
+
+            behavior.Invoke(context, () => { });
+
+            var routingStrategy = (ToAllSubscribers)context.Get<RoutingStrategy>();
```

Comment:
Did we agree whether we always do a Get<> or should there be a property? Don't remember sorry